v-next/hardhat/src/internal/builtin-plugins/artifacts/artifact-manager.ts

```diff
@@ -0,0 +1,405 @@
+      (p) => p.endsWith(".json") && !p.endsWith(".output.json"),
+    );
+
+    return new Set(paths.map((p) => path.basename(p, ".json")));
+  }
+
+  public async getBuildInfoPath(buildInfoId: string): Promise<string> {
+    return path.join(
+      this.#artifactsPath,
+      BUILD_INFO_DIR_NAME,
+      buildInfoId + ".json",
+    );
+  }
+
+  public async getBuildInfoOutputPath(
+    buildInfoId: string,
+  ): Promise<string | undefined> {
+    return path.join(
+      this.#artifactsPath,
+      BUILD_INFO_DIR_NAME,
+      buildInfoId + ".output.json",
+    );
+  }
+
+  public async clearCache(): Promise<void> {
+    this.#fsData = undefined;
+  }
+
+  async #getFullyQualifiedName(
+    contractNameOrFullyQualifiedName: string,
+  ): Promise<string> {
+    const { bareNameToFullyQualifiedNameMap, allFullyQualifiedNames } =
+      await this.#getFsData();
+
+    if (this.#isFullyQualifiedName(contractNameOrFullyQualifiedName)) {
+      if (allFullyQualifiedNames.has(contractNameOrFullyQualifiedName)) {
+        return contractNameOrFullyQualifiedName;
+      }
+
+      this.#throwNotFoundError(
+        contractNameOrFullyQualifiedName,
+        bareNameToFullyQualifiedNameMap,
+      );
+    }
+
+    const fqns = bareNameToFullyQualifiedNameMap.get(
+      contractNameOrFullyQualifiedName,
+    );
+
+    if (fqns === undefined || fqns.size === 0) {
+      this.#throwNotFoundError(
+        contractNameOrFullyQualifiedName,
+        bareNameToFullyQualifiedNameMap,
+      );
+    }
+
+    if (fqns.size !== 1) {
+      throw new HardhatError(HardhatError.ERRORS.ARTIFACTS.MULTIPLE_FOUND, {
+        contractName: contractNameOrFullyQualifiedName,
+        candidates: Array.from(fqns).join(EOL),
+      });
+    }
+
+    const [fqn] = fqns;
+
+    return fqn;
+  }
+
+  #throwNotFoundError(
+    contractNameOrFullyQualifiedName: string,
+    bareNameToFullyQualifiedNameMap: Map<string, ReadonlySet<string>>,
+  ): never {
+    const similarNames = this.#getSimilarContractNames(
+      contractNameOrFullyQualifiedName,
+      bareNameToFullyQualifiedNameMap,
+    );
+
+    const suggestion = this.#formatSimilarNameSuggestions(
+      contractNameOrFullyQualifiedName,
+      similarNames,
+    );
+
+    throw new HardhatError(HardhatError.ERRORS.ARTIFACTS.NOT_FOUND, {
+      contractName: contractNameOrFullyQualifiedName,
+      suggestion,
+    });
+  }
+
+  #isFullyQualifiedName(name: string): boolean {
+    return name.includes(":");
+  }
+
+  #getFullyQualifiedNameFromArtifactAbsolutePath(artifactPath: string): string {
+    const relativePath = path.relative(this.#artifactsPath, artifactPath);
+    const sourceName = path.dirname(relativePath).split(path.sep).join("/");
+    const contractName = path.basename(relativePath, ".json");
+    return `${sourceName}:${contractName}`;
+  }
+
+  #getSimilarContractNames(
+    contractNameOrFullyQualifiedName: string,
+    bareNameToFullyQualifiedNameMap: Map<string, ReadonlySet<string>>,
+  ): string[] {
```

Comment:
I found it a little confusing that this function can also return similar fully qualified names. Would it make sense to split it in two? `getSimilarContractNames` and `getSimilarFullyQualifiedNames`?

Reply:
Great point! I changed into a slightly different approach in 4c8e874afaedd9e593050eb8678f9891a8c922c8, which I believe is simpler.

Reply:
Lovely! Yes, that's great!